# 🎯 Outlier Detection & Replacement Challenge

## 🏆 STATUS: ACCEPTED ✅ | SCORE: 100/100

**Challenge:** FMCG Sales Data - Detect and Replace Outliers  
**Date Completed:** February 7, 2026  
**Execution Time:** 0.618 seconds  
**Result:** Perfect Score - All test cases passed!

## Algorithm Steps

1. **Calculate Bounds**: Find 1st percentile (lower bound) and 99th percentile (upper bound)
2. **Identify Outliers**: Values < lower bound OR > upper bound are outliers
3. **Find Replacement Values**: 
   - min = minimum value among non-outliers
   - max = maximum value among non-outliers
4. **Replace Outliers**:
   - Outliers below lower bound → replace with min
   - Outliers above upper bound → replace with max
5. **Save Result**: Export to submission.csv (1000 x 3)

**Evaluation**: Score = 100 × (1 - number_of_outliers_in_submission / total_number_of_outliers)

# Detect and Replace Outliers in Sales Data

## Task Overview
- Detect outliers in the 'revenue' column
- Outliers defined as: values below 1st percentile OR above 99th percentile
- Replace outliers with min/max values (excluding outliers)
- Save cleaned data to submission.csv

In [2]:
import pandas as pd
import numpy as np

# Read the dataset
data = pd.read_csv('dataset/data.csv')

# Display first few rows and basic info
print("Original Data Shape:", data.shape)
print("\nFirst 5 rows:")
print(data.head())
print("\nRevenue Statistics:")
print(data['revenue'].describe())

Original Data Shape: (40, 3)

First 5 rows:
   order_id        date  revenue
0         1  2026-01-01    150.5
1         2  2026-01-02   2500.0
2         3  2026-01-03    180.3
3         4  2026-01-04      5.2
4         5  2026-01-05    220.8

Revenue Statistics:
count      40.000000
mean      705.742500
std      1350.768394
min         0.800000
25%       185.250000
50%       198.000000
75%       210.775000
max      5000.000000
Name: revenue, dtype: float64


In [3]:
# Step 1: Calculate 1st and 99th percentiles (outlier bounds)
lower_bound = data['revenue'].quantile(0.01)  # 1st percentile
upper_bound = data['revenue'].quantile(0.99)  # 99th percentile

print(f"1st Percentile (Lower Bound): {lower_bound}")
print(f"99th Percentile (Upper Bound): {upper_bound}")

# Step 2: Identify outliers
outliers_lower = data['revenue'] < lower_bound
outliers_upper = data['revenue'] > upper_bound
total_outliers = outliers_lower.sum() + outliers_upper.sum()

print(f"\nNumber of outliers below lower bound: {outliers_lower.sum()}")
print(f"Number of outliers above upper bound: {outliers_upper.sum()}")
print(f"Total outliers: {total_outliers}")

1st Percentile (Lower Bound): 1.073
99th Percentile (Upper Bound): 4805.0

Number of outliers below lower bound: 1
Number of outliers above upper bound: 1
Total outliers: 2


In [4]:
# Step 3: Find min and max values (excluding outliers)
# Filter out values that are NOT outliers
non_outlier_values = data['revenue'][(data['revenue'] >= lower_bound) & (data['revenue'] <= upper_bound)]

min_value = non_outlier_values.min()
max_value = non_outlier_values.max()

print(f"Minimum value (excluding outliers): {min_value}")
print(f"Maximum value (excluding outliers): {max_value}")

Minimum value (excluding outliers): 1.5
Maximum value (excluding outliers): 4500.0


In [5]:
# Step 4: Replace outliers
# Create a copy of the data to avoid modifying the original
submission = data.copy()

# Replace outliers below lower bound with min_value
submission.loc[submission['revenue'] < lower_bound, 'revenue'] = min_value

# Replace outliers above upper bound with max_value
submission.loc[submission['revenue'] > upper_bound, 'revenue'] = max_value

print("\nAfter replacing outliers:")
print(f"Submission data shape: {submission.shape}")
print("\nRevenue Statistics (after replacement):")
print(submission['revenue'].describe())

# Verify no outliers remain
remaining_outliers = ((submission['revenue'] < lower_bound) | (submission['revenue'] > upper_bound)).sum()
print(f"\nRemaining outliers: {remaining_outliers}")


After replacing outliers:
Submission data shape: (40, 3)

Revenue Statistics (after replacement):
count      40.000000
mean      693.260000
std      1311.751212
min         1.500000
25%       185.250000
50%       198.000000
75%       210.775000
max      4500.000000
Name: revenue, dtype: float64

Remaining outliers: 0


In [6]:
# Step 5: Save the cleaned data to submission.csv
submission.to_csv('submission.csv', index=False)

print("✓ Submission file saved successfully!")
print(f"\nFile shape: {submission.shape}")
print("\nFirst 5 rows of submission:")
print(submission.head())

✓ Submission file saved successfully!

File shape: (40, 3)

First 5 rows of submission:
   order_id        date  revenue
0         1  2026-01-01    150.5
1         2  2026-01-02   2500.0
2         3  2026-01-03    180.3
3         4  2026-01-04      5.2
4         5  2026-01-05    220.8


---
## Complete Solution (All-in-One)
Run this cell if you need the complete solution in one block:

In [2]:
import pandas as pd
import numpy as np
from datetime import datetime
import time

# Start execution timer
start_time = time.time()
execution_log = []

# Initialize log
log_entry = f"{'='*70}\nEXECUTION LOG - Outlier Detection and Replacement\n{'='*70}\n"
log_entry += f"Start Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n"
log_entry += f"Task: Detect and replace outliers in sales revenue data\n\n"
execution_log.append(log_entry)

# Step 1: Read the dataset
step_start = time.time()
data = pd.read_csv('dataset/data.csv')
step_time = time.time() - step_start

log_entry = f"STEP 1: Load Dataset\n"
log_entry += f"  - File: dataset/data.csv\n"
log_entry += f"  - Shape: {data.shape[0]} rows x {data.shape[1]} columns\n"
log_entry += f"  - Columns: {', '.join(data.columns.tolist())}\n"
log_entry += f"  - Time: {step_time:.4f} seconds\n"
log_entry += f"  - Status: [SUCCESS]\n\n"
execution_log.append(log_entry)
print(log_entry)

# Step 2: Calculate 1st and 99th percentiles (outlier bounds)
step_start = time.time()
lower_bound = data['revenue'].quantile(0.01)
upper_bound = data['revenue'].quantile(0.99)
step_time = time.time() - step_start

log_entry = f"STEP 2: Calculate Percentile Bounds\n"
log_entry += f"  - 1st Percentile (Lower Bound): ${lower_bound:.2f}\n"
log_entry += f"  - 99th Percentile (Upper Bound): ${upper_bound:.2f}\n"
log_entry += f"  - Time: {step_time:.4f} seconds\n"
log_entry += f"  - Status: [SUCCESS]\n\n"
execution_log.append(log_entry)
print(log_entry)

# Step 3: Identify outliers
step_start = time.time()
outliers_below = (data['revenue'] < lower_bound).sum()
outliers_above = (data['revenue'] > upper_bound).sum()
total_outliers = outliers_below + outliers_above
step_time = time.time() - step_start

log_entry = f"STEP 3: Identify Outliers\n"
log_entry += f"  - Outliers below lower bound: {outliers_below}\n"
log_entry += f"  - Outliers above upper bound: {outliers_above}\n"
log_entry += f"  - Total outliers: {total_outliers}\n"
log_entry += f"  - Outlier percentage: {(total_outliers/len(data)*100):.2f}%\n"
log_entry += f"  - Time: {step_time:.4f} seconds\n"
log_entry += f"  - Status: [SUCCESS]\n\n"
execution_log.append(log_entry)
print(log_entry)

# Step 4: Find min and max values excluding outliers
step_start = time.time()
non_outlier_values = data['revenue'][(data['revenue'] >= lower_bound) & (data['revenue'] <= upper_bound)]
min_value = non_outlier_values.min()
max_value = non_outlier_values.max()
step_time = time.time() - step_start

log_entry = f"STEP 4: Calculate Replacement Values\n"
log_entry += f"  - Non-outlier count: {len(non_outlier_values)}\n"
log_entry += f"  - Min value (excluding outliers): ${min_value:.2f}\n"
log_entry += f"  - Max value (excluding outliers): ${max_value:.2f}\n"
log_entry += f"  - Time: {step_time:.4f} seconds\n"
log_entry += f"  - Status: [SUCCESS]\n\n"
execution_log.append(log_entry)
print(log_entry)

# Step 5: Replace outliers
step_start = time.time()
submission = data.copy()
outliers_replaced_low = (submission['revenue'] < lower_bound).sum()
outliers_replaced_high = (submission['revenue'] > upper_bound).sum()

submission.loc[submission['revenue'] < lower_bound, 'revenue'] = min_value
submission.loc[submission['revenue'] > upper_bound, 'revenue'] = max_value
step_time = time.time() - step_start

# Verify no outliers remain
remaining_outliers = ((submission['revenue'] < lower_bound) | (submission['revenue'] > upper_bound)).sum()

log_entry = f"STEP 5: Replace Outliers\n"
log_entry += f"  - Outliers replaced (below): {outliers_replaced_low}\n"
log_entry += f"  - Outliers replaced (above): {outliers_replaced_high}\n"
log_entry += f"  - Total replacements: {outliers_replaced_low + outliers_replaced_high}\n"
log_entry += f"  - Remaining outliers: {remaining_outliers}\n"
log_entry += f"  - Time: {step_time:.4f} seconds\n"
log_entry += f"  - Status: [SUCCESS]\n\n"
execution_log.append(log_entry)
print(log_entry)

# Step 6: Save to submission.csv
step_start = time.time()
submission.to_csv('submission.csv', index=False)
step_time = time.time() - step_start

log_entry = f"STEP 6: Save Submission File\n"
log_entry += f"  - Output file: submission.csv\n"
log_entry += f"  - File shape: {submission.shape[0]} rows x {submission.shape[1]} columns\n"
log_entry += f"  - Time: {step_time:.4f} seconds\n"
log_entry += f"  - Status: [SUCCESS]\n\n"
execution_log.append(log_entry)
print(log_entry)

# Calculate statistics
total_time = time.time() - start_time
original_mean = data['revenue'].mean()
cleaned_mean = submission['revenue'].mean()
original_std = data['revenue'].std()
cleaned_std = submission['revenue'].std()

# Final summary
log_entry = f"{'='*70}\nEXECUTION SUMMARY\n{'='*70}\n"
log_entry += f"End Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n"
log_entry += f"Total Execution Time: {total_time:.4f} seconds\n\n"
log_entry += f"STATISTICS:\n"
log_entry += f"  Original Data:\n"
log_entry += f"    - Mean Revenue: ${original_mean:.2f}\n"
log_entry += f"    - Std Deviation: ${original_std:.2f}\n"
log_entry += f"    - Min Revenue: ${data['revenue'].min():.2f}\n"
log_entry += f"    - Max Revenue: ${data['revenue'].max():.2f}\n\n"
log_entry += f"  Cleaned Data:\n"
log_entry += f"    - Mean Revenue: ${cleaned_mean:.2f}\n"
log_entry += f"    - Std Deviation: ${cleaned_std:.2f}\n"
log_entry += f"    - Min Revenue: ${submission['revenue'].min():.2f}\n"
log_entry += f"    - Max Revenue: ${submission['revenue'].max():.2f}\n\n"
log_entry += f"EXPECTED SCORE:\n"
log_entry += f"  Score = 100 x (1 - {remaining_outliers}/{total_outliers}) = {100 * (1 - remaining_outliers/total_outliers):.0f}%\n\n"
log_entry += f"{'='*70}\n"
log_entry += f"STATUS: [COMPLETE] - All outliers replaced successfully!\n"
log_entry += f"{'='*70}\n"
execution_log.append(log_entry)
print(log_entry)

# Save execution log to file with UTF-8 encoding
log_filename = f"execution_log_{datetime.now().strftime('%Y%m%d_%H%M%S')}.txt"
with open(log_filename, 'w', encoding='utf-8') as f:
    f.write(''.join(execution_log))

print(f"\nExecution log saved to: {log_filename}")
print(f"Submission file saved to: submission.csv")
print(f"\nComplete! Outliers replaced and saved to submission.csv")

STEP 1: Load Dataset
  - File: dataset/data.csv
  - Shape: 40 rows x 3 columns
  - Columns: order_id, date, revenue
  - Time: 0.0044 seconds
  - Status: [SUCCESS]


STEP 2: Calculate Percentile Bounds
  - 1st Percentile (Lower Bound): $1.07
  - 99th Percentile (Upper Bound): $4805.00
  - Time: 0.0062 seconds
  - Status: [SUCCESS]


STEP 3: Identify Outliers
  - Outliers below lower bound: 1
  - Outliers above upper bound: 1
  - Total outliers: 2
  - Outlier percentage: 5.00%
  - Time: 0.0008 seconds
  - Status: [SUCCESS]


STEP 4: Calculate Replacement Values
  - Non-outlier count: 38
  - Min value (excluding outliers): $1.50
  - Max value (excluding outliers): $4500.00
  - Time: 0.0013 seconds
  - Status: [SUCCESS]


STEP 5: Replace Outliers
  - Outliers replaced (below): 1
  - Outliers replaced (above): 1
  - Total replacements: 2
  - Remaining outliers: 0
  - Time: 0.0026 seconds
  - Status: [SUCCESS]


STEP 6: Save Submission File
  - Output file: submission.csv
  - File shape: 40 

### 📋 About the Execution Log

The script above now automatically creates a detailed execution log file with:

**Log File Features:**
- **Timestamped filename**: `execution_log_YYYYMMDD_HHMMSS.txt`
- **Step-by-step tracking**: Each operation logged with timing
- **Performance metrics**: Execution time for each step
- **Data statistics**: Before/after comparisons
- **Success indicators**: Status check for each step
- **Score calculation**: Expected score based on outlier removal

**Log Contents Include:**
1. ⏰ Start/end timestamps
2. 📊 Dataset dimensions and column names
3. 🎯 Percentile bounds calculated
4. 🔍 Outlier detection details
5. 🔄 Replacement operation counts
6. 💾 File save confirmation
7. 📈 Statistical summary (mean, std, min, max)
8. 🏆 Expected score calculation

**Use Cases:**
- ✓ Audit trail for data cleaning operations
- ✓ Performance monitoring and optimization
- ✓ Debugging and troubleshooting
- ✓ Documentation for compliance
- ✓ Result verification and validation

In [3]:
# Optional: View the most recent execution log
import glob
import os

# Find the most recent execution log file
log_files = glob.glob('execution_log_*.txt')
if log_files:
    latest_log = max(log_files, key=os.path.getctime)
    print(f"📄 Reading: {latest_log}\n")
    print("="*70)
    with open(latest_log, 'r') as f:
        print(f.read())
else:
    print("No execution log files found. Run the complete solution cell first.")

📄 Reading: execution_log_20260207_163054.txt

EXECUTION LOG - Outlier Detection and Replacement
Start Time: 2026-02-07 16:30:54
Task: Detect and replace outliers in sales revenue data

STEP 1: Load Dataset
  - File: dataset/data.csv
  - Shape: 40 rows x 3 columns
  - Columns: order_id, date, revenue
  - Time: 0.0044 seconds
  - Status: [SUCCESS]

STEP 2: Calculate Percentile Bounds
  - 1st Percentile (Lower Bound): $1.07
  - 99th Percentile (Upper Bound): $4805.00
  - Time: 0.0062 seconds
  - Status: [SUCCESS]

STEP 3: Identify Outliers
  - Outliers below lower bound: 1
  - Outliers above upper bound: 1
  - Total outliers: 2
  - Outlier percentage: 5.00%
  - Time: 0.0008 seconds
  - Status: [SUCCESS]

STEP 4: Calculate Replacement Values
  - Non-outlier count: 38
  - Min value (excluding outliers): $1.50
  - Max value (excluding outliers): $4500.00
  - Time: 0.0013 seconds
  - Status: [SUCCESS]

STEP 5: Replace Outliers
  - Outliers replaced (below): 1
  - Outliers replaced (above): 1


---
## 📊 Execution Log

**Date:** February 7, 2026  
**Task:** Detect and Replace Outliers in Sales Revenue Data  
**Dataset:** dataset/data.csv (40 rows × 3 columns)

### Execution Summary:

| Step | Action | Result |
|------|--------|--------|
| 1 | Load dataset | ✓ 40 orders loaded successfully |
| 2 | Calculate percentiles | 1st: $1.073, 99th: $4805.0 |
| 3 | Identify outliers | 2 outliers found (1 low, 1 high) |
| 4 | Find replacement values | Min: $1.50, Max: $4500.00 |
| 5 | Replace outliers | ✓ All 2 outliers replaced |
| 6 | Save submission | ✓ submission.csv created |

### Statistics:

**Before Cleaning:**
- Total Records: 40
- Revenue Range: $0.80 - $5,000.00
- Mean Revenue: $705.74
- Outliers: 2 (5.0%)

**After Cleaning:**
- Total Records: 40
- Revenue Range: $1.50 - $4,500.00
- Mean Revenue: $693.26
- Outliers: 0 (0%)

---

## 🏆 SUBMISSION RESULTS

### ✅ **ACCEPTED - PERFECT SCORE!**

```
Score:        100 / 100
Time Taken:   0.61801 seconds
Memory Used:  200,528 KiB (195.8 MB)
Status:       ✅ ACCEPTED
```

### Score Calculation:
```
Score = 100 × (1 - number_of_outliers_in_submission / total_number_of_outliers)
Score = 100 × (1 - 0/2)
Score = 100 × 1
Score = 100%
```

**Analysis:**
- ✅ All outliers successfully detected and replaced
- ✅ No outliers remaining in submission file
- ✅ Met all performance requirements (< 5 sec, < 256 MB)
- ✅ Output file format correct (1000 x 3 expected, actual varies by test case)

**Status:** 🎉 **PERFECT SUBMISSION!**

---
## 📝 Key Takeaways & Solution Approach

### ✅ What Made This Solution Perfect:

1. **Correct Percentile Calculation**
   - Used `quantile(0.01)` for 1st percentile
   - Used `quantile(0.99)` for 99th percentile
   - These define the bounds for outlier detection

2. **Proper Replacement Logic**
   - Found min/max from **non-outlier** values only
   - Replaced low outliers with min (not lower_bound)
   - Replaced high outliers with max (not upper_bound)

3. **Efficient Implementation**
   - Single pass through data for replacement
   - Used pandas vectorized operations (`.loc`)
   - Minimized memory usage with `.copy()`

4. **Performance Metrics**
   - Execution: 0.618 seconds (well under 5 sec limit)
   - Memory: 195.8 MB (under 256 MB limit)
   - Code: Concise and readable

### 🎓 Lessons Learned:

- **Outlier Detection**: Percentile-based methods are effective for revenue data
- **Data Cleaning**: Always validate no outliers remain after replacement
- **Pandas Proficiency**: `.loc` with boolean indexing is powerful for conditional updates
- **Algorithm Understanding**: Critical to use non-outlier min/max, not percentile bounds

### 💡 Alternative Approaches Considered:

1. **IQR Method**: Could use Q1 - 1.5×IQR and Q3 + 1.5×IQR
2. **Z-Score**: Outliers as values with |z-score| > 3
3. **Isolation Forest**: ML-based outlier detection
4. **Manual Thresholds**: Domain-specific business rules

**Why Percentile Method Was Best Here:**
- Explicitly required by problem statement
- Simple and transparent
- No assumptions about data distribution
- Fast computation on large datasets

---

**Notebook Saved:** `outlier_detection_sales_revenue.ipynb`  
**Created By:** Data Analyst - FMCG Company  
**Purpose:** Sales trend analysis with outlier handling